builtin/providers/google/resource_container_cluster.go

```diff
@@ -0,0 +1,383 @@
+	if len(masterAuths) > 1 {
+		return fmt.Errorf("Cannot specify more than one master_auth.")
+	}
+	masterAuth := masterAuths[0].(map[string]interface{})
+
+	cluster := &container.Cluster{
+		MasterAuth: &container.MasterAuth{
+			Password: masterAuth["password"].(string),
+			User:     masterAuth["username"].(string),
+		},
+		Name:     clusterName,
+		NumNodes: int64(d.Get("number_of_nodes").(int)),
+	}
+
+	if v, ok := masterAuth["bearer_token"]; ok {
+		cluster.MasterAuth.BearerToken = v.(string)
+	}
+
+	if v, ok := d.GetOk("cluster_api_version"); ok {
+		cluster.ClusterApiVersion = v.(string)
+	}
+
+	if v, ok := d.GetOk("container_ipv4_cidr"); ok {
+		cluster.ContainerIpv4Cidr = v.(string)
+	}
+
+	if v, ok := d.GetOk("description"); ok {
+		cluster.Description = v.(string)
+	}
+
+	if v, ok := d.GetOk("enable_cloud_logging"); ok {
+		cluster.EnableCloudLogging = v.(bool)
+	}
+
+	if v, ok := d.GetOk("network_name"); ok {
+		cluster.Network = v.(string)
+	}
+
+	if v, ok := d.GetOk("node_config"); ok {
+		nodeConfigs := v.(*schema.Set).List()
+		if len(nodeConfigs) > 1 {
+			return fmt.Errorf("Cannot specify more than one node_config.")
+		}
+		nodeConfig := nodeConfigs[0].(map[string]interface{})
+
+		cluster.NodeConfig = &container.NodeConfig{}
+
+		if v, ok = nodeConfig["machine_type"]; ok {
+			cluster.NodeConfig.MachineType = v.(string)
+		}
+		if v, ok = nodeConfig["source_image"]; ok {
+			cluster.NodeConfig.SourceImage = v.(string)
+		}
+
+		if v, ok := nodeConfig["service_accounts"]; ok {
+			accounts := v.(*schema.Set).List()
+
+			serviceAccounts := []*container.ServiceAccount{}
+			for _, acc := range accounts {
+				account := acc.(map[string]interface{})
+				scopesList := account["scopes"].([]interface{})
+				scopes := []string{}
+				for _, v := range scopesList {
+					scopes = append(scopes, v.(string))
+				}
+
+				serviceAccounts = append(serviceAccounts, &container.ServiceAccount{
+					Email:  account["email"].(string),
+					Scopes: scopes,
+				})
+			}
+
+			cluster.NodeConfig.ServiceAccounts = serviceAccounts
+		}
+	}
+
+	req := &container.CreateClusterRequest{
+		Cluster: cluster,
+	}
+
+	op, err := config.clientContainer.Projects.Zones.Clusters.Create(
+		config.Project, zoneName, req).Do()
+	if err != nil {
+		return err
+	}
+
+	// Wait until it's created
+	wait := resource.StateChangeConf{
+		Pending:    []string{"pending", "running"},
+		Target:     "done",
+		Timeout:    30 * time.Minute,
+		MinTimeout: 3 * time.Second,
+		Refresh: func() (interface{}, string, error) {
+			resp, err := config.clientContainer.Projects.Zones.Operations.Get(
+				config.Project, zoneName, op.Name).Do()
+			log.Printf("[DEBUG] Progress of creating GKE cluster %s: %s",
+				clusterName, resp.Status)
+			return resp, resp.Status, err
+		},
+	}
+
+	_, err = wait.WaitForState()
+	if err != nil {
+		return err
+	}
+
+	log.Printf("[INFO] GKE cluster %s has been created", clusterName)
+
+	d.SetId(zoneName + "-" + clusterName)
```

Comment:
I was thinking that it was necessary due to the way zones work in the API - e.g. I can create 3 GKE clusters using the same `provider`, each in different zone. But now I realise that Terraform is actually abstracting that problem away by `resource "" "unique_name" { ...` so it can be just `clusterName`.

Good catch!